cql3/statements/alter_keyspace_statement.cc

```diff
@@ -110,6 +113,35 @@ void cql3::statements::alter_keyspace_statement::validate(query_processor& qp, c
                         throw exceptions::invalid_request_exception("Only one DC's RF can be changed at a time and not by more than 1");
                     }
                 }
+
+                const auto& rack_map = tmptr->get_topology().get_datacenter_racks();
+
+                // We cannot allow for ALTER to proceed if it would lead to an RF-rack-invalid keyspace.
+                // For more context, see: scylladb/scylladb#23276.
+                for (const auto& [new_dc, rf] : new_rf_per_dc) {
+                    const auto rf_value = parse_rf(rf);
+                    size_t normal_rack_count = 0;
+                    for (const auto& [_, rack_nodes] : rack_map.at(new_dc)) {
+                        // We must ignore zero-token nodes because they don't take part in replication.
+                        // Verify that this rack has at least one normal node.
+                        const bool normal_rack = std::ranges::any_of(rack_nodes, [tmptr] (locator::host_id host_id) {
+                            return tmptr->is_normal_token_owner(host_id);
+                        });
+                        if (normal_rack) {
+                            ++normal_rack_count;
+                        }
+                    }
+                    const bool restricted_keyspaces = qp.db().get_config().rf_rack_valid_keyspaces();
+                    const bool invalid_rf = (rf_value != (int) normal_rack_count) && rf_value != 1 && rf_value != 0;
+
+                    if (restricted_keyspaces && invalid_rf) {
```

Comment:
It may be fine for now as a guardrail, but it is racy. If there is a concurrent topology change, the validation check may be invalid by the time we actually apply the RF change in the transition handler. So we should validate the invariants there as well. I understand that this is left for later.